src/path-util.typ

```diff
@@ -123,37 +127,91 @@
 /// - segments (array): List of path segments
 /// -> float: Total length of the path
 #let length(segments) = {
-  return segments.map(segment-length).sum()
+  return segments.map(_segment-length).sum()
+}
+
+/// Normalize path by inserting linear segments
+/// if the path contains gaps
+///
+/// - segments (array): List of path segments
+/// - close (bool): If true, insert a close segment if the first and last
+///   path point are different
+///
+/// -> array List of new segments
+#let normalize-path(segments, close: false) = {
+  let new = ()
+
+  let prev-pt = none
+  for s in segments {
+    let kind = s.at(0)
+    let start = segment-start(s)
+    let end = segment-end(s)
+
+    if prev-pt != none and prev-pt != start {
+      new.push(("line", prev-pt, start))
+    }
+
+    new.push(s)
+    prev-pt = end
+  }
+
+  if close and segment-start(new.first()) != segment-end(new.last()) {
+    new.push(("line", segment-start(new.first()), segment-end(new.last())))
+  }
+
+  return new
 }
 
 /// Get position on path
 ///
 /// - segments (array): List of path segments
-/// - t (float): Position (from 0 to 1)
+/// - t (float,ratio): Position (from 0 to 1)
 /// -> vector: Position on path
 #let point-on-path(segments, t) = {
-  if segments.len() == 1 {
-    return point-on-segment(segments.first(), t)
-  }
+  assert(type(t) in (int, float, ratio),
+    message: "Distance t must be of type number or ratio")
 
-  let l = length(segments)
+  if segments.len() == 0 {
+    return none
+  } else if segments.len() == 1 {
+    return _point-on-segment(segments.first(), t)
+  }
 
-  let traveled-length = 0
-  for s in segments {
-    let part-length = segment-length(s)
+  let abs = type(t) != ratio
+  let target = calc.max(0, if type(t) == ratio {
+    t * length(segments) / 100%
+  } else {
+    t
+  })
```

Comment:
We should check that `t` is in an acceptable range before this, instead of snapping to the start of the segment. There also doesn't seem to be a check for if the distance is greater than the length of the path.